# Exploring Hackers News Posts

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

In this project, we'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

In [1]:
# Read in the data.
import csv

f = open('hacker_news.csv')
hn = list(csv.reader(f))
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

# Removing Headers from a List of Lists

In [2]:
headers = hn[0]
hn = hn[1:]
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [3]:
hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

# Extracting Ask HN and Show HN Posts
First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [4]:
ask_posts = []
show_posts = []
other_posts = []

for x in hn:
    title1 = x[1]
    if title1.lower().startswith('ask hn'):
        ask_posts.append(x)
    elif title1.lower().startswith('show hn'):
        show_posts.append(x)
    else:
        other_posts.append(x)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


# Calculating the Average Number of Comments for Ask HN and Show HN Posts
Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [5]:
print(headers)
print(ask_posts[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [6]:
total_ask_comments = 0
for x in ask_posts:
    num_comments = int(x[4])
    total_ask_comments += num_comments

avg_ask_comments = (total_ask_comments / len(ask_posts))
print('Avg No. of comments on Ask Posts:',avg_ask_comments)

Avg No. of comments on Ask Posts: 14.038417431192661


In [7]:
total_show_comments = 0
for x in show_posts:
    num_comments = int(x[4])
    total_show_comments += num_comments

avg_show_comments = (total_show_comments / len(show_posts))
print('Avg No. of comments on Ask Posts:',avg_show_comments)

Avg No. of comments on Ask Posts: 10.31669535283993


Ask Posts(14) receives more comments than the show posts(10) and hence we'll do more analysis on the ask_post

In [8]:
print(ask_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [9]:
import datetime as dt

In [10]:
result_list = []

In [11]:
for post in ask_posts:
    result_list.append([post[6], int(post[4])])

In [12]:
print(result_list[6])

['4/22/2016 12:24', 4]


# Finding the Amount of Ask Posts and Comments by Hour Created
Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [13]:
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for x in result_list:
    date = x[0]
    time = dt.datetime.strptime(date, date_format)
    date_1 = time.strftime("%H")
    
    if date_1 in counts_by_hour:
        counts_by_hour[date_1] += 1
        comments_by_hour[date_1] += x[1]
    else:
        counts_by_hour[date_1] = 1
        comments_by_hour[date_1] = x[1]
        
comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

# Calculating the Average Number of Comments for Ask HN Posts by Hour

In [14]:
avg_by_hour = []

for hr in counts_by_hour:
    if hr in comments_by_hour:
        avg_by_hour.append([hr,(comments_by_hour[hr]/counts_by_hour[hr])])
avg_by_hour


[['13', 14.741176470588234],
 ['07', 7.852941176470588],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['12', 9.41095890410959],
 ['11', 11.051724137931034],
 ['03', 7.796296296296297],
 ['20', 21.525],
 ['08', 10.25],
 ['06', 9.022727272727273],
 ['04', 7.170212765957447],
 ['01', 11.383333333333333],
 ['09', 5.5777777777777775],
 ['02', 23.810344827586206],
 ['23', 7.985294117647059],
 ['22', 6.746478873239437],
 ['21', 16.009174311926607],
 ['15', 38.5948275862069],
 ['18', 13.20183486238532],
 ['00', 8.127272727272727],
 ['10', 13.440677966101696],
 ['19', 10.8],
 ['17', 11.46],
 ['05', 10.08695652173913]]

# Sorting and Printing Values from a List of Lists

In [17]:
swap_avg_by_hour = []

for hr in avg_by_hour:
    swap_avg_by_hour.append([hr[1],hr[0]])
    
swap_avg_by_hour
    

[[14.741176470588234, '13'],
 [7.852941176470588, '07'],
 [13.233644859813085, '14'],
 [16.796296296296298, '16'],
 [9.41095890410959, '12'],
 [11.051724137931034, '11'],
 [7.796296296296297, '03'],
 [21.525, '20'],
 [10.25, '08'],
 [9.022727272727273, '06'],
 [7.170212765957447, '04'],
 [11.383333333333333, '01'],
 [5.5777777777777775, '09'],
 [23.810344827586206, '02'],
 [7.985294117647059, '23'],
 [6.746478873239437, '22'],
 [16.009174311926607, '21'],
 [38.5948275862069, '15'],
 [13.20183486238532, '18'],
 [8.127272727272727, '00'],
 [13.440677966101696, '10'],
 [10.8, '19'],
 [11.46, '17'],
 [10.08695652173913, '05']]

In [18]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

In [19]:
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

# Printing top 5 hours for Ask Posts Comments

In [21]:
print('Top 5 Hours for Ask Posts Comments')
date_format = "%H"

for x in sorted_swap[:5]:
    date = x[1]
    date1 = dt.datetime.strptime(date, date_format).strftime('%H:%M')
    str_format = "{}: {:.2f} average comments per post"
    print(str_format.format(date1,x[0]))
    

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.